components/public-api/gitpod/v1/workspace.proto

```diff
@@ -319,17 +319,64 @@ message WorkspaceMetadata {
 
   // warnings are user-facing warnings that should be displayed to the user when trying to start the workspace
   repeated string warnings = 8;
+
+  // context is the context from which the workspace is created
+  WorkspaceContext context = 9;
+}
+
+// WorkspaceContext is the git context from which the workspace is created
+message WorkspaceContext {
```

Comment:
This was added so that we can fill in the values for `contextURLSegment_1`,  `contextURLSegment_2` and `contextURL_cloneURL` in the CSV. I imagine that it can come in handy to get it directly from the API as well.

What I want to draw attention to is that I did not add the ws ctx as a top-level property (like it is for workspaces in the DB and in the protocol), but rather hid it under `metadata`. I'm happy to push it up, but it made sense at the time I was trying to use it.

Reply:
While I think what you proposed makes sense in itself, and feels like a nice "completion" of our API, I don't feel the usecase we have is a compelling reason to add the `WorkspaceContext` shape (with all it's complexity) to the (central) `Workspace` shape.

My two main concerns are:
1. extent: most of the fields exposed here we won't use going forward - not even in the narrow use-case we have right now
2. exposing the concept of `WorkspaceContext` right now, with such little benefit, but a significant "cost"
  - this shows in the "where to place it" problem
  - adding it now would severely reduce our solution space if we really needed it in the future (a bit speculative given the state of Classic, but still)

Could we get a way without adding this shape at all? E.g. treat contextUrl as URL, and just focus on the path elements? :thinking:

Reply:
@geropl this is definitely fair. The addition of the workspace context is something I tried to really get around - it's true we most likely won't be able to use this much outside of the context of workspace sessions and insights[^1].

I [have tried](https://github.com/gitpod-io/gitpod/pull/20437/commits/100c003c19308f150e9260cdc4c38b884339f0a0#diff-6f589748220cc1e35424b542adef97af969bb8e6d8080309629c38048588ec75R182-R184) some simple parsing of context URLs, but because we don't offer the `cloneURL` on the Workspace responses, there is close to no way to parse URLs and split segments properly without running into the question of "is this a repo name part, a subgroup or already some extra path like `/commits`?".

Additionally, we could really expose some bare-bones information, potentially just on the `WorkspaceSession` object.

[^1]: that said, the flip side here is that now folks utilizing our `ListWorkspaceSessions` responses will be able to have a lot more data about the contexts, and will be able to answer adoption-related questions like "do users open PRs/MRs with Gitpod?"

Reply:
> that said, the flip side here is that now folks utilizing our ListWorkspaceSessions responses will be able to have a lot more data about the contexts, and will be able to answer adoption-related questions like "do users open PRs/MRs with Gitpod?"

That is a good argument indeed. Let me dwell a bit on it.
 
> I [have tried](https://github.com/gitpod-io/gitpod/pull/20437/commits/100c003c19308f150e9260cdc4c38b884339f0a0#diff-6f589748220cc1e35424b542adef97af969bb8e6d8080309629c38048588ec75R182-R184) some simple parsing of context URLs, but because we don't offer the cloneURL

Argh, you are right. The naming of fields is really all over the place :grimacing:

Reply:
In sync, we decided to keep all of the details, but stick them under the `WorkspaceSession` shape.

Reply:
Please ping me once it's finalized and the preview is ready (still seeing the "react flower" on the bottom, so you might still be changing things...?). :pray: